File: other-programming-languages/Perl/2024/2024-010/2024-010.pl
use strict;
use warnings;
use Time::HiRes qw(gettimeofday tv_interval);

my $INPUT_FILE = "input.txt";

sub read_map {
    my ($filename) = @_;
    open my $fh, '<', $filename or die "Cannot open file: $!";
    my @grid;
    while (my $line = <$fh>) {
        chomp $line;
        push @grid, [map { int($_) } split //, $line];
    }
    close $fh;
    return \@grid;
}

sub neighbors {
    my ($r, $c, $rows, $cols) = @_;
    my @directions = ([-1, 0], [1, 0], [0, -1], [0, 1]);
    my @result;
    for my $dir (@directions) {
        my ($dr, $dc) = @$dir;
        my $nr = $r + $dr;
        my $nc = $c + $dc;
        if ($nr >= 0 && $nr < $rows && $nc >= 0 && $nc < $cols) {
            push @result, [$nr, $nc];
        }
    }
    return @result;
}

sub find_trailhead_scores {
    my ($grid) = @_;
    my $rows = @$grid;
    my $cols = @{$grid->[0]};

    my @trailheads;
    for my $r (0 .. $rows - 1) {
        for my $c (0 .. $cols - 1) {
            push @trailheads, [$r, $c] if $grid->[$r][$c] == 0;
        }
    }

    my $total_score = 0;

    for my $start (@trailheads) {
        my ($start_r, $start_c) = @$start;
        my %visited;
        my @queue = ([$start_r, $start_c]);
        $visited{"$start_r,$start_c"} = 1;

        my %reachable_nines;

        while (@queue) {
            my ($r, $c) = @{shift @queue};
            my $current_height = $grid->[$r][$c];

            if ($current_height == 9) {
                $reachable_nines{"$r,$c"} = 1;
            } else {
                my $next_height = $current_height + 1;
                for my $neighbor (neighbors($r, $c, $rows, $cols)) {
                    my ($nr, $nc) = @$neighbor;
                    if (!$visited{"$nr,$nc"} && $grid->[$nr][$nc] == $next_height) {
                        $visited{"$nr,$nc"} = 1;
                        push @queue, [$nr, $nc];
                    }
                }
            }
        }

        $total_score += keys %reachable_nines;
    }

    return $total_score;
}

sub count_paths {
    my ($r, $c, $grid, $dp, $rows, $cols) = @_;
    return $dp->[$r][$c] if defined $dp->[$r][$c];

    my $current_height = $grid->[$r][$c];

    if ($current_height == 9) {
        $dp->[$r][$c] = 1;
        return 1;
    }

    my $total_paths = 0;
    my $next_height = $current_height + 1;
    for my $neighbor (neighbors($r, $c, $rows, $cols)) {
        my ($nr, $nc) = @$neighbor;
        if ($grid->[$nr][$nc] == $next_height) {
            $total_paths += count_paths($nr, $nc, $grid, $dp, $rows, $cols);
        }
    }

    $dp->[$r][$c] = $total_paths;
    return $total_paths;
}

sub calculate_total_rating {
    my ($grid) = @_;
    my $rows = @$grid;
    my $cols = @{$grid->[0]};

    my @trailheads;
    for my $r (0 .. $rows - 1) {
        for my $c (0 .. $cols - 1) {
            push @trailheads, [$r, $c] if $grid->[$r][$c] == 0;
        }
    }

    my @dp = map { [(undef) x $cols] } 0 .. $rows - 1;

    my $total_rating = 0;
    for my $start (@trailheads) {
        my ($tr, $tc) = @$start;
        $total_rating += count_paths($tr, $tc, $grid, \@dp, $rows, $cols);
    }

    return $total_rating;
}

sub main {
    my $start_time_part1 = [gettimeofday];
    my $grid = read_map($INPUT_FILE);
    my $total_score = find_trailhead_scores($grid);
    my $elapsed_time_part1 = tv_interval($start_time_part1);

    print "Part 1 Result: $total_score\n";
    printf "Time taken for Part 1: %.9f s\n", $elapsed_time_part1;

    my $start_time_part2 = [gettimeofday];
    my $total_rating = calculate_total_rating($grid);
    my $elapsed_time_part2 = tv_interval($start_time_part2);

    print "Part 2 Result: $total_rating\n";
    printf "Time taken for Part 2: %.9f s\n", $elapsed_time_part2;
}

main();

File: other-programming-languages/Go/2024/2024-010/2024-010.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

const INPUT_FILE = `input.txt`

func readMap(filename string) [][]int {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var grid [][]int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if len(line) > 0 {
			row := []int{}
			for _, ch := range line {
				num, _ := strconv.Atoi(string(ch))
				row = append(row, num)
			}
			grid = append(grid, row)
		}
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	return grid
}

func neighbors(r, c, rows, cols int) [][2]int {
	directions := [][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
	var result [][2]int
	for _, dir := range directions {
		nr, nc := r+dir[0], c+dir[1]
		if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
			result = append(result, [2]int{nr, nc})
		}
	}
	return result
}

func findTrailheadScores(grid [][]int) int {
	rows := len(grid)
	cols := len(grid[0])

	var trailheads [][2]int
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] == 0 {
				trailheads = append(trailheads, [2]int{r, c})
			}
		}
	}

	totalScore := 0

	for _, start := range trailheads {
		startR, startC := start[0], start[1]

		visited := map[[2]int]bool{}
		queue := [][2]int{{startR, startC}}
		visited[[2]int{startR, startC}] = true

		reachableNines := map[[2]int]bool{}

		for len(queue) > 0 {
			r, c := queue[0][0], queue[0][1]
			queue = queue[1:]

			currentHeight := grid[r][c]

			if currentHeight == 9 {
				reachableNines[[2]int{r, c}] = true
			} else {
				nextHeight := currentHeight + 1
				for _, neighbor := range neighbors(r, c, rows, cols) {
					nr, nc := neighbor[0], neighbor[1]
					if !visited[[2]int{nr, nc}] && grid[nr][nc] == nextHeight {
						visited[[2]int{nr, nc}] = true
						queue = append(queue, [2]int{nr, nc})
					}
				}
			}
		}

		totalScore += len(reachableNines)
	}

	return totalScore
}

func countPaths(r, c int, grid [][]int, dp [][]int, rows, cols int) int {
	if dp[r][c] != -1 {
		return dp[r][c]
	}

	currentHeight := grid[r][c]

	if currentHeight == 9 {
		dp[r][c] = 1
		return 1
	}

	totalPaths := 0
	nextHeight := currentHeight + 1
	for _, neighbor := range neighbors(r, c, rows, cols) {
		nr, nc := neighbor[0], neighbor[1]
		if grid[nr][nc] == nextHeight {
			totalPaths += countPaths(nr, nc, grid, dp, rows, cols)
		}
	}

	dp[r][c] = totalPaths
	return totalPaths
}

func calculateTotalRating(grid [][]int) int {
	rows := len(grid)
	cols := len(grid[0])

	var trailheads [][2]int
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] == 0 {
				trailheads = append(trailheads, [2]int{r, c})
			}
		}
	}

	dp := make([][]int, rows)
	for i := range dp {
		dp[i] = make([]int, cols)
		for j := range dp[i] {
			dp[i][j] = -1
		}
	}

	totalRating := 0
	for _, start := range trailheads {
		tr, tc := start[0], start[1]
		totalRating += countPaths(tr, tc, grid, dp, rows, cols)
	}

	return totalRating
}

func main() {
	startTimePart1 := time.Now()
	grid := readMap(INPUT_FILE)
	totalScore := findTrailheadScores(grid)
	endTimePart1 := time.Now()
	fmt.Printf("Part 1 Result: %d\n", totalScore)
	fmt.Printf("Time taken for Part 1: %.9f s\n", endTimePart1.Sub(startTimePart1).Seconds())

	startTimePart2 := time.Now()
	totalRating := calculateTotalRating(grid)
	endTimePart2 := time.Now()
	fmt.Printf("Part 2 Result: %d\n", totalRating)
	fmt.Printf("Time taken for Part 2: %.9f s\n", endTimePart2.Sub(startTimePart2).Seconds())
}
